### PR TITLE
Added support for BER decoding of indefinite length fields

### PR DESCRIPTION
**Pull request by Finndersen**
A project I'm working on requires decoding of indefinite-length fields for BER encoding. 
I made required changes to support it and appears to be working well.
Added Specification.decode_with_length() method which returns decoded data and byte length, useful for indefinite-length decoding.
Tested with indefinite length Sets and Sequences

### CODE REVIEW

**Comment by eerimoq** [comment]
First of all, revert all whitespace changes. There is no reason to change the coding style.

Second, I highly doubt this works as you implemented very few tests. Please add more tests.

**Comment by Finndersen** [comment]
Fixed that up.
It works for telecommunication network records BER encoded with indefinite length constructed SETS, SEQUENCE and CHOICEs.
I've created tests for those constructed types.
Is there any more you would like to see?

**Comment by Finndersen** [comment]
Sorry missed tests for decode_with_length(), doing that now

File: asn1tools/codecs/ber.py [hunk 1]
```diff
@@ -473,7 +476,6 @@ def decode_constructed_segments(self, segments):
 
 
 class StringType(PrimitiveOrConstructedType):
-
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:476]: Revert.

File: asn1tools/codecs/ber.py [hunk 2]
```diff
@@ -570,26 +572,37 @@ def decode(self, data, offset):
         offset = self.decode_tag(data, offset)
 
         if data[offset] == 0x80:
-            raise NotImplementedError(
-                'Decode until an end-of-contents tag is not yet implemented.')
+            # Indefinite length field
+            offset = offset + 1
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:576]: offset += 1

File: asn1tools/codecs/ber.py [hunk 3]
```diff
@@ -570,26 +572,37 @@ def decode(self, data, offset):
...
+            end_offset = None
+            # raise NotImplementedError('Decode until an end-of-contents tag is not yet implemented.')
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:578]: Remove.

File: asn1tools/codecs/ber.py [hunk 4]
```diff
@@ -570,26 +572,37 @@ def decode(self, data, offset):
...
         if self.additions:
-            self.decode_additions(data,
-                                  values,
-                                  offset,
-                                  end_offset)
+            offset = self.decode_additions(data,
+                                           values,
+                                           offset,
+                                           end_offset)
+        # Detect end of indefinite length constructed field
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:600]: Add a blank line before the comment.
End the commend with a period (``.``).

File: asn1tools/codecs/ber.py [hunk 5]
```diff
@@ -570,26 +572,37 @@ def decode(self, data, offset):
...
+        if end_offset is None:
+            if data[offset:offset + 2] == EOC_TAG:
+                end_offset = offset + 2
+            else:
+                raise DecodeError('Could not find end-of-contents tag for indefinite length field')
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:605]: End the sentence with a period.

File: asn1tools/codecs/ber.py [hunk 6]
```diff
@@ -615,16 +628,20 @@ def decode_additions(self, data, values, offset, end_offset):
                 values.update(addition_values)
         except DecodeError:
             pass
+        return offset
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:631]: Add a blknk line before the return.

File: asn1tools/codecs/ber.py [hunk 7]
```diff
@@ -615,16 +628,20 @@ def decode_additions(self, data, values, offset, end_offset):
...
+            # If reached end of indefinite length field, decode should raise DecodeTagError,
+            # and end of field will be handled in .decode() method
+            if end_offset is None or offset < end_offset:
                 if isinstance(member, AnyDefinedBy):
                     value, offset = member.decode(data, offset, values)
                 else:
                     value, offset = member.decode(data, offset)
             else:
                 raise IndexError
+
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:644]: Revert.

File: asn1tools/codecs/ber.py [hunk 8]
```diff
@@ -674,17 +691,25 @@ def encode(self, data, encoded):
 
     def decode(self, data, offset):
         offset = self.decode_tag(data, offset)
-
         if data[offset] == 0x80:
-            raise NotImplementedError(
-                'Decode until an end-of-contents tag is not yet implemented.')
+            offset = offset + 1
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:695]: offset += 1

File: asn1tools/codecs/ber.py [hunk 9]
```diff
@@ -674,17 +691,25 @@ def encode(self, data, encoded):
...
+            length = None  # Indicates indefinite field
+            # raise NotImplementedError('Decode until an end-of-contents tag is not yet implemented.')
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:697]: Remove the comment.

File: asn1tools/codecs/ber.py [hunk 10]
```diff
@@ -674,17 +691,25 @@ def encode(self, data, encoded):
...
         else:
             length, offset = decode_length_definite(data, offset)
 
         decoded = []
         start_offset = offset
-
-        while (offset - start_offset) < length:
+        # Loop through data until length exceeded or end-of-contents tag reached
+        while 1:
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:704]: while True

File: asn1tools/codecs/ber.py [hunk 11]
```diff
@@ -1381,9 +1406,21 @@ def encode(self, data, encoded):
 
     def decode(self, data, offset):
         offset = self.decode_tag(data, offset)
-        _, offset = decode_length_definite(data, offset)
+        indefinite = False
+        if data[offset] == 0x80:
+            # Indefinite length field
+            offset = offset + 1
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:1412]: offset += 1

File: asn1tools/codecs/ber.py [hunk 12]
```diff
@@ -1381,9 +1406,21 @@ def encode(self, data, encoded):
 
     def decode(self, data, offset):
         offset = self.decode_tag(data, offset)
-        _, offset = decode_length_definite(data, offset)
+        indefinite = False
+        if data[offset] == 0x80:
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:1410]: Add a blank line before if statement.

File: asn1tools/codecs/ber.py [hunk 13]
```diff
@@ -1381,9 +1406,21 @@ def encode(self, data, encoded):
...
+            # Indefinite length field
+            offset = offset + 1
+            indefinite = True
+        else:
+            # Definite length field
+            length, offset = decode_length_definite(data, offset)
 
-        return self.inner.decode(data, offset)
+        values, end_offset = self.inner.decode(data, offset)
+        if indefinite:
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:1419]: Add a blank line before the if statement.

File: asn1tools/codecs/ber.py [hunk 14]
```diff
@@ -1381,9 +1406,21 @@ def encode(self, data, encoded):
...
+            if data[end_offset:end_offset+2] != EOC_TAG:
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:1420]: ``data[end_offset:end_offset + 2]``

File: asn1tools/codecs/ber.py [hunk 15]
```diff
@@ -1381,9 +1406,21 @@ def encode(self, data, encoded):
...
+                raise DecodeError('Expected End-Of-Contents tag at offset: {}'.format(end_offset))
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:1421]: Inconsistency. "End-Of-Contents" was "end-of-contents" in other exception in the same file. Use lower case?

File: asn1tools/codecs/ber.py [hunk 16]
```diff
@@ -1381,9 +1406,21 @@ def encode(self, data, encoded):
...
+            end_offset += 2
+        return values, end_offset
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:1423]: Add a blank line before return.

**Comment by Finndersen** [comment]
Addressed

**Comment by eerimoq** [comment]
Great!

The test may of course be sufficient to enure good quality as most of the code is shared with the definite length variant. If you are confident it works, fine. =)

File: asn1tools/codecs/ber.py [hunk 17]
```diff
@@ -71,6 +71,9 @@ class Tag(object):
     DATE_TIME         = 0x21
 
 
+EOC_TAG = bytes(2)
```
Comment (eerimoq) [inline asn1tools/codecs/ber.py:74]: Hmmm, I think I prefer ``END_OF_CONTENTS_OCTETS`` to make it easier to understand

**Comment by Finndersen** [comment]
I've tested it with real-world indefinite-length BER encoded telecommunications records.
And passes other tests as expected. 
Could potentially be some problematic edge cases since I'm only familiar with this field of encoding styles.
In any case, it shouldn't break any existing functionality

File: tests/test_ber.py [hunk 18]
```diff
@@ -717,6 +726,17 @@ def test_set(self):
         for type_name, decoded, encoded in datas:
             self.assert_encode_decode(foo, type_name, decoded, encoded)
 
+        # Test indefinite length set with end-of-contents tags
+        self.assertEqual(foo.decode('A', b'\x31\x80\x80\x01\x03\x81\x01\x00\x00\x00'), {'a': 3, 'b': 0})
+        self.assertEqual(foo.decode('C', b'\x31\x80\xa0\x80\x80\x01\x03\x81\x01\x04\x00\x00\x81\x01\x12\x00\x00'), {'a': {'a': 3, 'b': 4}, 'b': b'\x12'})
```
Comment (eerimoq) [inline tests/test_ber.py:731]: Aim for maximum line length of 85 characters.

File: asn1tools/compiler.py [hunk 19]
```diff
@@ -167,6 +167,42 @@ def decode(self, name, data, check_constraints=False):
...
+        """
+
+        try:
+            type_ = self._types[name]
+        except KeyError:
+            raise DecodeError(
+                "Type '{}' not found in types dictionary.".format(name))
+
+        if not hasattr(type_, 'decode_with_length'):
+            raise DecodeError("Type '{}' does not support decode_with_length".format(name))
```
Comment (eerimoq) [inline asn1tools/compiler.py:197]: Make all types have this method. Those that does not implement it should return ``NotImplementedError(a message)``

File: asn1tools/compiler.py [hunk 20]
```diff
@@ -167,6 +167,42 @@ def decode(self, name, data, check_constraints=False):
...
+        ConstraintsError exception is raised if the constraints are
+        not fulfilled. Set `check_constraints` to ``False`` to skip
+        the constraints check and minimize the runtime overhead, but
+        instead allow decoding of values not fulfilling the
+        constraints.
+
+        >>> foo.decode_with_length('Question', b'0\\x0e\\x02\\x01\\x01\\x16\\x09Is 1+1=3?')
+        ({'id': 1, 'question': 'Is 1+1=3?'}, 16)
+
+        """
```
Comment (eerimoq) [inline asn1tools/compiler.py:188]: This docstring is mostly a copy from the decode function. Add a refernece to the decode function, and wexaplin the differences here.

File: asn1tools/compiler.py (hunk 19 above)
Comment (Finndersen) [reply asn1tools/compiler.py:197]: I was going to add abstract decode_with_length() to base Type class but it doesn't even have abstract .decode() method?
And will a top-level type of BER Specification always be a CompiledType (wrapping another type)?

Comment (eerimoq) [reply asn1tools/compiler.py:197]: Maybe I forgot to add the decode method. It's a mistake in that case.

I think so, yes. It was quite some time since I worked on this project, so I might be wrong.

File: asn1tools/compiler.py (hunk 20 above)
Comment (eerimoq) [reply asn1tools/compiler.py:188]: I didn't mean add this commect to the review. Most of the two docstrings are not the same, only the constraints part.

File: asn1tools/compiler.py (hunk 19 above)
Comment (Finndersen) [reply asn1tools/compiler.py:197]: Base Type for all codecs do not have abstract .decode() method.
Also CompileType behaves differently to standard Type (acts as entry point for decoding?), and its .encode() bahaves differently (just calls wrapped type .decode() and strips the returned offset)
So I don't think it makes sense to add decode_with_length() to all Types (Types already return data and offset)
I can add abstract decode_with_length() to base CompiledType class, however it also does not have an abstract .decode()

Comment (eerimoq) [reply asn1tools/compiler.py:197]: Yeah it should be added to the wrapper class CompiledType. Try to add it and implement a test calling it for a codec that does not implement it. If the expected exception is raised we are ok.

Comment (Finndersen) [reply asn1tools/compiler.py:197]: Looks like all codec implementations of CompiledType share a common pattern (setting self._type on __init__, type() property, __repr__ implementation), looks like this should be moved to base CompiledType class? along with abstract .decode() and .decode_with_length()

Comment (Finndersen) [reply asn1tools/compiler.py]: Yep done

**Comment by eerimoq** [comment]
Very nice job, thank you!

For some reason the Travis CI build result is not shown in this PR. See https://travis-ci.org/github/eerimoq/asn1tools/jobs/694642995 for failed Python 2 tests.

I guess support for Python 2 should be dropped soon, but lets' make it work in this PR.

**Comment by Finndersen** [comment]
That should fix it
